---
title: "DifferentialAnalysesPigBAL"
author: "Arianne Richard"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Intro

This script picks up where "clustering all.Rmd" left off. 

We will start by labelling the clusters and doing some filtering to narrow to those clusters that we intend to run differential analyses on. We will then run differential analyses between conditions. Specifically, we will be looking for differential abundance of cells within each cluster, as well as differential expression of cells within each cluster, according to treatment condition. There are 4 conditions: PBS, Ad-HA/NP, Ad-HA/NP+Ad-IL1b, pH1N1. There are male and female mice for each group, but sex is almost completely confounded with sets of runs, so we are unable to account for both of these in the same model. We will address this problem in both differential abundance and differential expression analyses. 

# Loading the data

```{r}
library(SingleCellExperiment)
library(DropletUtils)
library(scran)
library(scater)
library(uwot)
library(biomaRt)
library(scds)
library(batchelor)
library(bluster)
library(gridExtra)
library(edgeR)

load('ongoing_full_clustered.RData')

## size of saved data got a bit out of hand
rm(sce_list)

```

# Cluster annotation

Based on manual annotation from the clustered data performed using the generated tables and our shiny app, we'll give the clusters more intuitive names that can be transposed into figures. As a quick check in this script, we look at particular marker genes first.

```{r}
## first look at marker genes for sorting

sce_all$cluster <- sce_all_corrected$cluster

plotExpression(sce_all_corrected, features=rownames(sce_all)[rowData(sce_all)$Symbol == 'CD4'], x='cluster', exprs_values='logcounts')
plotExpression(sce_all, features=rownames(sce_all)[rowData(sce_all)$Symbol == 'CD4'], x='cluster')

plotExpression(sce_all_corrected, features=rownames(sce_all)[rowData(sce_all)$Symbol == 'CD8B'], x='cluster', exprs_values='logcounts')
plotExpression(sce_all, features=rownames(sce_all)[rowData(sce_all)$Symbol == 'CD8B'], x='cluster')

plotExpression(sce_all_corrected, features=rownames(sce_all)[rowData(sce_all)$Symbol == 'CD163'], x='cluster', exprs_values='logcounts')
plotExpression(sce_all, features=rownames(sce_all)[rowData(sce_all)$Symbol == 'CD163'], x='cluster')

plotExpression(sce_all_corrected, features=rownames(sce_all)[rowData(sce_all)$Symbol == 'CD14'], x='cluster', exprs_values='logcounts')
plotExpression(sce_all, features=rownames(sce_all)[rowData(sce_all)$Symbol == 'CD14'], x='cluster')

plotExpression(sce_all_corrected, features=rownames(sce_all)[rowData(sce_all)$Symbol == 'CD3D'], x='cluster', exprs_values='logcounts')
plotExpression(sce_all, features=rownames(sce_all)[rowData(sce_all)$Symbol == 'CD3D'], x='cluster')

plotExpression(sce_all_corrected, features=rownames(sce_all)[rowData(sce_all)$Symbol == 'CD3E'], x='cluster', exprs_values='logcounts')
plotExpression(sce_all, features=rownames(sce_all)[rowData(sce_all)$Symbol == 'CD3E'], x='cluster')

```

Based on our examinations in the clustering script, we exclude cluster 5 as low quality and clusters 17 and 11 as doublets. 

Note that we are maintaining 2 separate datasets here: sce_filt (from sce_all_corrected), which contains 5000 HVGs and reconstructed values after batch correction; and sce_all_filt (from sce_all), which contains all detected genes with normalised (but not batch corrected) log-count values and original counts. sce_filt is important for visualization (e.g. UMAP), whereas sce_all_filt is important for count-based differential analyses and seeing expression values that aren't interpolated after batch correction.

```{r}

sce_filt <- sce_all_corrected[,!sce_all_corrected$cluster %in% c('5', '17', '11')]

sce_all_filt <- sce_all[,!sce_all$cluster %in% c('5', '17', '11')]
```

Next we name the good quality clusters. 

```{r}
## add in cluster names as sce_all_filt$named_cluster
#Long-form annotated version including cell subtypes
for (i in unique(sce_all_filt$cluster)){
  if(any(i==c('7', '20a'))){
    name <- 'CD4 T Cells' 
    long_name <- 'CD4 T Cells, 7, 20a'}
  if(any(i==c('18b'))){
    name <- 'Mitotic CD4 T Cells'
    long_name <- 'Mitotic CD4 T Cells, 18b'}
  if(any(i==c('10'))){
    name <- 'Tregs'
    long_name <- 'Tregs, 10'}
  if(any(i==c('6', '9a', '20b'))){
    name <- 'CD8 T Cells' 
    long_name <- 'CD8 T Cells, 6, 9a, 20b'}
  if(any(i==c('18a'))){
    name <- 'Mitotic CD8 T Cells'
    long_name <- 'Mitotic CD8 T Cells, 18a'}
  if(any(i==c('1', '4', '8'))){
    name <- 'Macrophages' 
    long_name <- 'Macrophages, 1, 4, 8'}
  if(any(i==c('15'))){
    name <- 'Monocytes'
    long_name <- 'Monocytes, 15'}
  if(any(i==c('13'))){
    name <- 'cDCs'
    long_name <- 'cDCs, 13'}
  if(any(i==c('2', '3'))){
    name <- 'Plasma Cells' 
    long_name <- 'Plasma Cells, 2, 3'}
  if(any(i==c('12', '16'))){
    name <- 'B Cells' 
    long_name <- 'B Cells, 12, 16'}
  if(any(i==c('14'))){
    name <- 'T-cell-like' 
    long_name <- 'T-cell-like, 14'}
  if(any(i==c('9b', '9c'))){
    name <- 'NK cells' 
    long_name <- 'NK cells, 9b, 9c'}
  if(any(i==c('21'))){
    name <- 'Unknown'
    long_name <- 'Unknown, 21'}
  if(any(i==c('19'))){
    name <- 'Epithelial cells' 
    long_name <- 'Epithelial Cells, 19'}
  sce_all_filt$named_cluster[sce_all_filt$cluster==i] <- name
  sce_all_filt$long_named_cluster[sce_all_filt$cluster==i] <- long_name
}

# create 'main_clusters' column that ignores subclustering
main_clusters <- gsub("a", "", sce_all_filt$cluster)
main_clusters <- gsub("b", "", main_clusters)
main_clusters <- gsub("c", "", main_clusters)
sce_all_filt$main_clusters <- main_clusters

# Create label and colour slots for convenient annotation later
sce_all_filt$verbose_cluster <- paste0('BAL, ', sce_all_filt$named_cluster, ', ', sce_all_filt$cluster)
sce_all_filt$EXP_celltype <- paste0('BAL, ', sce_all_filt$named_cluster)

BAL_colour_list <- vector("list", length(sce_all_filt$EXP_celltype))
BAL_colour_list[grep('Macrophages', sce_all_filt$EXP_celltype)] <- '#E31A1C' #red
BAL_colour_list[grep('Monocytes', sce_all_filt$EXP_celltype)] <- '#FB9A99' #pink
BAL_colour_list[grep('cDCs', sce_all_filt$EXP_celltype)] <- '#6A3D9A' #purple
BAL_colour_list[grep('Plasma Cells', sce_all_filt$EXP_celltype)] <- '#1F78B4' #dark blue
BAL_colour_list[grep('B Cells', sce_all_filt$EXP_celltype)] <- '#A6CEE3' #pale blue
BAL_colour_list[grep('Tregs', sce_all_filt$EXP_celltype)] <- '#CAB2D6' #light purple
BAL_colour_list[grep('CD4', sce_all_filt$EXP_celltype)] <- '#B2DF8A' #light green
BAL_colour_list[grep('Mitotic CD4 T Cells', sce_all_filt$EXP_celltype)] <- '#FDBF6F' #pale orange
BAL_colour_list[grep('CD8', sce_all_filt$EXP_celltype)] <- '#33A02C' # green
BAL_colour_list[grep('Mitotic CD8 T Cells', sce_all_filt$EXP_celltype)] <- '#FF7F00' #orange
BAL_colour_list[grep('NK', sce_all_filt$EXP_celltype)] <- '#FFFF00' #yellow
BAL_colour_list[sapply(BAL_colour_list, is.null)] <- '#B15928' #brown
sce_all_filt$colour  <- BAL_colour_list

colData(sce_filt)[,c('named_cluster', 'long_named_cluster', 'main_clusters', 'verbose_cluster', 'EXP_celltype', 'colour')] <- colData(sce_all_filt)[, c('named_cluster', 'long_named_cluster', 'main_clusters', 'verbose_cluster', 'EXP_celltype', 'colour')]

```

And make some UMAPs.
```{r}
legend_colours <- BAL_colour_list
names(legend_colours) <- sce_all_filt$long_named_cluster
legend_colours <- legend_colours[!duplicated(names(legend_colours))]
legend_colours <- legend_colours[c("Macrophages, 1, 4, 8", "Monocytes, 15", "cDCs, 13", "Plasma Cells, 2, 3", "B Cells, 12, 16", "CD4 T Cells, 7, 20a", "Tregs, 10", "Mitotic CD4 T Cells, 18b", "Mitotic CD8 T Cells, 18a", "CD8 T Cells, 6, 9a, 20b", "NK cells, 9b, 9c", "T-cell-like, 14", "Epithelial Cells, 19", "Unknown, 21")]

df_legend_colours <- data.frame(unlist(names(legend_colours)), unlist(legend_colours))
names(df_legend_colours) <- c("Cell_Type", "Colour")
df_legend_colours$Cell_Type <- factor(df_legend_colours$Cell_Type, levels=c(df_legend_colours$Cell_Type))
df_legend_colours$dummy_data <- seq(1, length(df_legend_colours[,1]))

# Plot new named clusters
png('Results/UMAP_named_clusters.png', height=1080, width=2160)
p1 <- plotUMAP(sce_filt, colour_by="cluster", point_size=0.3) + guides(colour = guide_legend(override.aes = list(size=8)))+ theme(text = element_text(size = 20))
p2 <- plotUMAP(sce_filt, colour_by="named_cluster", point_size=0.3) + guides(colour = guide_legend(override.aes = list(size=8)))+ theme(text = element_text(size = 20))
grid.arrange(p1, p2, ncol=2)
dev.off()

png(filename='Results/UMAP_named_and_numbered_clusters.png', height=1080, width=2160)
plot0 <- ggplot(data=df_legend_colours, aes(x=dummy_data)) +
  geom_point(aes(y=dummy_data, color=Cell_Type), size=10) +
  labs(x=NULL, y=NULL, color=NULL) +
  scale_color_manual(values = legend_colours) +
  guides(colour = guide_legend(ncol = 1)) + 
  theme(legend.position=c(0.5,0), legend.direction="vertical", aspect.ratio=0, text = element_text(size = 50),
        axis.text.x=element_blank(), axis.ticks.x=element_blank(),
        axis.text.y=element_blank(), axis.ticks.y=element_blank())
plot1 <- plotUMAP(sce_filt, point_size=0, text_by="main_clusters", text_size=14) +
  aes(colour = sce_filt$colour) +
  geom_point(shape='.', size=1, alpha=0.5) +
  geom_text(x=-10, y=14, size=30, label="C", color="black") +
  theme(aspect.ratio = 1, plot.margin = unit(c(0,0,0,0), "cm"), text = element_text(size = 40))
grid.arrange(plot1, plot0, ncol=2)
dev.off()

png(filename='Results/UMAP_named_and_unnumbered_clusters.png', height=1080, width=2160)
plot0 <- ggplot(data=df_legend_colours, aes(x=dummy_data)) +
  geom_point(aes(y=dummy_data, color=Cell_Type), size=10) +
  labs(x=NULL, y=NULL, color=NULL) +
  scale_color_manual(values = legend_colours) +
  guides(colour = guide_legend(ncol = 1)) + 
  theme(legend.position=c(0.5,0), legend.direction="vertical", aspect.ratio=0, text = element_text(size = 50),
        axis.text.x=element_blank(), axis.ticks.x=element_blank(),
        axis.text.y=element_blank(), axis.ticks.y=element_blank())
plot1 <- plotUMAP(sce_filt, point_size=0) +
  aes(colour = sce_filt$colour) +
  geom_point(shape='.', size=1, alpha=0.5) +
  geom_text(x=-10, y=14, size=30, label="C", color="black") +
  theme(aspect.ratio = 1, plot.margin = unit(c(0,0,0,0), "cm"), text = element_text(size = 40))
grid.arrange(plot1, plot0, ncol=2)
dev.off()
```
investigate FOXP3 and Tregs
```{r}
library(viridis)
#FOXP3 = "ENSSSCG00000012300"
sce_FOXP3 <- sce_filt

names(rowData(sce_FOXP3))[which(names(rowData(sce_FOXP3))=="start")]="gene_start"
names(rowData(sce_FOXP3))[which(names(rowData(sce_FOXP3))=="end")]="gene_end"
range(as.vector(logcounts(sce_FOXP3['ENSSSCG00000012300' ,])))
range(as.vector(assay(sce_FOXP3['ENSSSCG00000012300' ,], "reconstructed")))

png(filename='Results/UMAP_FOXP3.png', height=2160, width=2160)
plotUMAP(sce_FOXP3, colour_by="ENSSSCG00000012300", point_size=0.3, by_exprs_values="reconstructed") + 
  theme(text = element_text(size = 20)) + 
  scale_color_gradientn(colours = rev(rocket(256)), limits=c(0,0.02))
dev.off()

# delete cells that show no FOXP3 expression

sce_FOXP3 <- sce_FOXP3[, assay(sce_FOXP3, "reconstructed")['ENSSSCG00000012300', ] > 0.001]

png(filename='Results/UMAP_FOXP3_thin.png', height=2160, width=2160)
plotUMAP(sce_FOXP3, colour_by="ENSSSCG00000012300", point_size=0.3, by_exprs_values="reconstructed") + 
  theme(text = element_text(size = 20)) + 
  scale_color_gradientn(colours = turbo(20), limits=c(0,0.01))
dev.off()

sce_FOXP3 <- sce_FOXP3[, assay(sce_FOXP3, "reconstructed")['ENSSSCG00000012300', ] > 0.002]

png(filename='Results/UMAP_FOXP3_superthin.png', height=2160, width=2160)
plotUMAP(sce_FOXP3, point_size=0) + 
  aes(colour = sce_FOXP3$colour) + 
  geom_point(shape='.', size=1, alpha=0.5) +
  theme(text = element_text(size = 20))
dev.off()

## all CD4+ and FOXP3+ cells as proportion of total cells (to mimic flow)
sce_filt$Flow_Treg[(sce_filt$cluster %in% c('10','7','18b','20a')) & (assay(sce_filt, "reconstructed")['ENSSSCG00000012300', ] > 0.002)] <- "YES"
sce_filt$Flow_Treg[is.na(sce_filt$Flow_Treg)] <- "NO"

CD4s <- sce_filt[,sce_filt$cluster %in% c('7', '10', '18b', '20a')]

props <- prop.table(table(sce_filt$Flow_Treg, sce_filt$Sample), margin=2)['YES',]
proptab <- annot[,c('sample', 'Condition', 'Sex')]
proptab$proportions <- props[rownames(proptab)]
proptab$Condition <- factor(proptab$Condition, levels=c('PBS', 'Ad-HA/NP', 'Ad-HA/NP+Ad-IL1b', 'pH1N1'))

ggplot(data=proptab, aes(x=Condition, y=proportions)) +
  geom_boxplot() +
  geom_point(aes(col=Sex), position='jitter')

## and print it

pdf('Results/CD4+FOXP3+_prop_total.pdf', height=10, width=10)
ggplot(data=proptab, aes(x=Condition, y=proportions)) +
  geom_boxplot(na.rm=TRUE) +
  geom_point(aes(col=Sex), position='jitter', size=4) +
  labs(x=NULL) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust=1), text=element_text(size=21))
dev.off()

#rm(sce_FOXP3)

```

And a TSNE for supplementary
```{r}
png(filename='Results/TSNE_named_and_unnumbered_clusters.png', height=1080, width=2160)
plot0 <- ggplot(data=df_legend_colours, aes(x=dummy_data)) +
  geom_point(aes(y=dummy_data, color=Cell_Type), size=10) +
  labs(x=NULL, y=NULL, color=NULL) +
  scale_color_manual(values = legend_colours) +
  guides(colour = guide_legend(ncol = 1)) + 
  theme(legend.position=c(0.5,0), legend.direction="vertical", aspect.ratio=0, text = element_text(size = 50),
        axis.text.x=element_blank(), axis.ticks.x=element_blank(),
        axis.text.y=element_blank(), axis.ticks.y=element_blank())
plot1 <- plotTSNE(sce_filt, point_size=0) +
  aes(colour = sce_filt$colour) +
  geom_point(size=1, alpha=1) +
  theme(aspect.ratio = 1, plot.margin = unit(c(0,0,0,0), "cm"), text = element_text(size = 40))
grid.arrange(plot1, plot0, ncol=2)
dev.off()
```


And a heatmap of all defining gene markers for our cell types
```{r}
library(gplots)
library(viridis)

defining_genes <- c("CD163", "FLT3", "CD14", "CD4", "CADM1", "CSF1R", "FCGR1A", "BCL6", "CCL17", "IRF8", "ADAM8", "PRDM1", "IRF4", "JCHAIN" , "TNFRSF17", "TNFRSF13B", "CD19", "PAX5", "EBF1", "MS4A1", "CD79B", "IRF8", "CD28", "CD5", "CD3D", "CD3E", "CD8A", "CD8B", "FOXP3", "E2F1", "E2F2", "KLRK1", "KLRB1", "KIT", "NCR1", "TBX21", "EOMES", "GATA3", "IL1RL1", "IL7R", "ID3", "ACTG1", "GPR37", "NAPSA", "SFTPA1", "RPS3A", "RPS26", "RPS11", "RPL37")

defining_sce <- sce_all_filt
rownames(defining_sce) <- rowData(sce_all_filt)$Symbol
# Remove forbidden rowData names (for generation of CompressedGRangesList)
names(rowData(defining_sce))[which(names(rowData(defining_sce))=="start")]="gene_start"
names(rowData(defining_sce))[which(names(rowData(defining_sce))=="end")]="gene_end"
#subset to defining genes and calculate mean logcounts
defining_sce <- defining_sce[rownames(defining_sce) %in% defining_genes, ]
defining_matrix <- as.matrix(logcounts(aggregateAcrossCells(defining_sce, ids=defining_sce$long_named_cluster, statistics='mean', use.assay.type='logcounts')))

xAxis <- colnames(defining_matrix)
#xColours <- ifelse(grepl("BAL", xAxis, fixed=TRUE), "blue", "orange")
colour_df <- unique(as.data.frame(colData(defining_sce))[, c("long_named_cluster", "colour")])
colour_df <- colour_df[match(xAxis, colour_df$long_named_cluster),]
identical(colour_df$long_named_cluster, xAxis)
Col_Colours <- as.character(colour_df$colour)

png(filename='Results/heatmap_defining_genes.png', height=2160, width=3840)
heatmap.2(defining_matrix, margins=c(20,20), scale="row", col = bluered(100), cexRow=2, cexCol=2,  trace="none", key=TRUE, key.xlab="row Z-score", lhei=c(1,10), lwid=c(1,10), ColSideColors=c(Col_Colours))
dev.off()

```
And a bubbleplot of defining genes
```{r}
#reorder axis's using heatmap dendrogram
pp <- heatmap.2(defining_matrix, scale="row")
rowIndex <- pp$rowInd
colIndex <- pp$colInd

png(filename='Results/bubbleplot_defining_genes.png', height=3840, width=3840)
plotDots(defining_sce, features=rownames(defining_sce), group="long_named_cluster", color=turbo(256)) +
  theme_minimal(base_size = 50) +
  theme(aspect.ratio = 6, axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.title.x=element_blank(), axis.title.y=element_blank(), legend.key.size = unit(2, 'cm')) +
  scale_y_discrete(limits=rownames(defining_matrix)[rowIndex]) +
  scale_x_discrete(limits=colnames(defining_matrix)[colIndex])
dev.off()
```


Based on the sorting strategy used, we did not intend to capture CD16+ cells (incl NK cells and NK-cell-like T cells) or epithelial cells. For this reason, we are reluctant to run differential analyses on clusters 9 (NK and NK-like CD8+ Ts) and cluster 19 (epithelial cells) as these are likely a biased set of cells within these phenotypes that snuck through the sorts. We will therefore exclude these clusters as well.

```{r}
sce_all_filt <- sce_all_filt[,!sce_all_filt$cluster %in% c('19', '9a', '9b', '9c')]
sce_filt <- sce_filt[,!sce_filt$cluster %in% c('19', '9a', '9b', '9c')]

legend_colours2 <- legend_colours[-c(grep('Epithelial', names(legend_colours)), grep('NK', names(legend_colours)))]

names(legend_colours2) <- c("Macrophages, 1, 4, 8", "Monocytes, 15", "cDCs, 13", "Plasma Cells, 2, 3", "B Cells, 12, 16", "CD4 T Cells, 7, 20a", "Tregs, 10", "Mitotic CD4 T Cells, 18b", "Mitotic CD8 T Cells, 18a", "CD8 T Cells, 6, 20b", "T-cell-like, 14", "Unknown, 21")

df_legend_colours2 <- data.frame(unlist(names(legend_colours2)), unlist(legend_colours2))
names(df_legend_colours2) <- c("Cell_Type", "Colour")
df_legend_colours2$Cell_Type <- factor(df_legend_colours2$Cell_Type, levels=c(df_legend_colours2$Cell_Type))
df_legend_colours2$dummy_data <- seq(1, length(df_legend_colours2[,1]))


png(filename='Results/UMAP_named_and_numbered_clusters_sorted_only.png', height=1080, width=2160)
plot0 <- ggplot(data=df_legend_colours2, aes(x=dummy_data)) +
  geom_point(aes(y=dummy_data, color=Cell_Type), size=10) +
  labs(x=NULL, y=NULL, color=NULL) +
  scale_color_manual(values = legend_colours2) +
  guides(colour = guide_legend(ncol = 1)) + 
  theme(legend.position=c(0.5,0), legend.direction="vertical", aspect.ratio=0, text = element_text(size = 50),
        axis.text.x=element_blank(), axis.ticks.x=element_blank(),
        axis.text.y=element_blank(), axis.ticks.y=element_blank())
plot1 <- plotUMAP(sce_filt, point_size=0, text_by="main_clusters", text_size=14) +
  aes(colour = sce_filt$colour) +
  geom_point(shape='.', size=1, alpha=0.5) +
  geom_text(x=-10, y=15, size=30, label="C", color="black") +
  theme(aspect.ratio = 1, plot.margin = unit(c(0,0,0,0), "cm"), text = element_text(size = 40))
grid.arrange(plot1, plot0, ncol=2)
dev.off()


```


# Differential abundance analysis

For running DA analysis, we must account for the sorting strategy that was used by only comparing clusters that would have fallen into the same sorting gate (a bit of a glorified FACS analysis really). For this reason, we will split the data into the sorted populations of macrophages/monocytes/DCs, CD4+ Ts, CD8b+ Ts, and B cells before looking for differential abundance within these groups. Of note, our differential abundance analyses will take place within categories that are smaller or the same size as those used for sorting. In this way, the sorting and binning won't play a role in our results.

CD4: 7, 10, 18b, 20a

CD8: 6, 18a, 20b

Macrophage/Monocyte/DC: 1, 4, 8, 13, 15

B: 2, 3, 12, 16

This leaves clusters 14 and 21 unconsidered as we are not completely sure how they were sorted.

After running each of the sorted groups separately, we'll also run a global differential abundance analysis that ignores the sorting strategy, just out of interest. In theory, if the same proportions of each sorted population were indeed pooled for sequencing, the results should be very similar and will give us a chance to look at the clusters that we were not able to assign to a sorted class.

```{r}

CD4s <- sce_all_filt[,sce_all_filt$cluster %in% c('7', '10', '18b', '20a')]

CD8s <- sce_all_filt[,sce_all_filt$cluster %in% c('6', '18a', '20b')]

MMDs <- sce_all_filt[,sce_all_filt$cluster %in% c('1', '4', '8', '13', '15')]

Bs <- sce_all_filt[,sce_all_filt$cluster %in% c('2', '3', '12', '16')]

all <- sce_all_filt

subsets <- list(CD4s=CD4s, CD8s=CD8s, MMDs=MMDs, Bs=Bs, all=all)

```

Our DA analysis will be run using a negative binomial test. Because clusters were computed on the dataset after merging and correcting for differences between runs, we will not include run as a covariate here. Instead, we will account for sex while testing differential abundance between conditions. 

We'll consider a change in abundance with any treatment versus PBS in an ANOVA-like analysis (strictly speaking, I believe this is analysis-of-deviance-like).

```{r}

DA <- lapply(subsets, function(x){
  abundances <- table(x$cluster, x$Sample)
  abundances <- unclass(abundances)
  anno <- annot[colnames(abundances),]
  y.ab <- DGEList(abundances, samples=anno)
  y.ab$samples$Sex <- factor(y.ab$samples$Sex)
  y.ab$samples$Condition <- factor(y.ab$samples$Condition, levels=c('PBS', 'Ad-HA/NP', 'Ad-HA/NP+Ad-IL1b', 'pH1N1'))
  design <- model.matrix(~Sex + Condition, y.ab$samples)
  y.ab <- estimateDisp(y.ab, design, trend="none")
  summary(y.ab$common.dispersion)
  plotBCV(y.ab, cex=1)
  fit.ab <- glmQLFit(y.ab, design, robust=TRUE, abundance.trend=FALSE)
  summary(fit.ab$var.prior)
  summary(fit.ab$df.prior)
  plotQLDisp(fit.ab, cex=1)
  res <- glmQLFTest(fit.ab, coef=3:5)
  restab <- res$table
  restab$FDR <- p.adjust(restab$PValue, method='fdr')
  return(restab)
})

DA
```

Interestingly, the only strong effect here is on the proportion of Treg in with vaccine + IL1b. Let's plot it to have a look.

```{r}
## proportion of total cells
props <- prop.table(table(all$cluster, all$Sample), margin=2)['10',]
proptab <- annot[,c('sample', 'Condition', 'Sex')]
proptab$proportions <- props[rownames(proptab)]
proptab$Condition <- factor(proptab$Condition, levels=c('PBS', 'Ad-HA/NP', 'Ad-HA/NP+Ad-IL1b', 'pH1N1'))

ggplot(data=proptab, aes(x=Condition, y=proportions)) +
  geom_boxplot() +
  geom_point(aes(col=Sex), position='jitter')

## and print it

pdf('Results/Treg_prop_total.pdf', height=10, width=10)
ggplot(data=proptab, aes(x=Condition, y=proportions)) +
  geom_boxplot(na.rm=TRUE) +
  geom_point(aes(col=Sex), position='jitter', size=4) +
  labs(x=NULL) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust=1), text=element_text(size=21))
dev.off()


## proportion of CD4s
props <- prop.table(table(CD4s$cluster, CD4s$Sample), margin=2)['10',]
proptab <- annot[,c('sample', 'Condition', 'Sex')]
proptab$proportions <- props[rownames(proptab)]
proptab$Condition <- factor(proptab$Condition, levels=c('PBS', 'Ad-HA/NP', 'Ad-HA/NP+Ad-IL1b', 'pH1N1'))

ggplot(data=proptab, aes(x=Condition, y=proportions)) +
  geom_boxplot() +
  geom_point(aes(col=Sex), position='jitter')

## and print it

pdf('Results/Treg_prop_CD4.pdf', height=10, width=10)
ggplot(data=proptab, aes(x=Condition, y=proportions)) +
  geom_boxplot(na.rm=TRUE) +
  geom_point(aes(col=Sex), position='jitter', size=4) +
  labs(x=NULL) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust=1), text=element_text(size=21))
dev.off()


## and another format for among CD4s
props <- prop.table(table(CD4s$cluster, CD4s$Sample), margin=2)
props <- data.frame(t(props))
colnames(props) <- c('sample', 'cluster', 'prop')
props$sample <- factor(props$sample, levels=c('sample1', 'sample8', 'sample11', 'sample2', 'sample12', 'sample5', 'sample6' ,'sample9', 'sample4', 'sample7', 'sample10'))

ggplot(props, aes(fill=cluster, y=prop, x=sample)) +
  geom_bar(position='stack', stat='identity')

## and print it
pdf('Results/CD4_cluster_proportions.pdf', height=10, width=10)
ggplot(props, aes(fill=cluster, y=prop, x=sample)) +
  geom_bar(position='stack', stat='identity') +
  labs(x=NULL, y='proportion') +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust=1), text=element_text(size=21))
dev.off()


```



# Differential expression analysis

Now we move on to differential expression analyses. Here, our decision about correcting for run versus sex is more complicated as we will be using the raw values which do exhibit run-specific effects. Comparing results from sex- versus run-corrected data revealed some Y-chromosome genes were being called as DE when we only corrected for run. This is a problem as the balance of M/F differs by condition (either 1v2 or 2v1). To avoid this, we'll have to correct for sex and accept that there is noise generated by each run that will simply reduce our power.

Now we test for differential expression. This analysis will be run as a pseudobulk analysis, pooling counts from all of the cells within each cluster for each sample before treating like bulk RNAseq data. Differential expression will then be analysed for each cluster. For this first pass, we will use an ANOVA-like analysis (again, more like ANODEV I believe) for each cluster to ask whether any condition deviates from the others. 

```{r}

summed <- aggregateAcrossCells(sce_all_filt, 
    id=colData(sce_filt)[,c("cluster", "Sample")])

summed.filt <- summed[,summed$ncells >= 10]

pseudoBulkDGEmod <- function(clust, sce, coeffs, filtexprsmin=20, filtexprstotal=40){
  sc <- sce[,sce$cluster==clust]
  y <- DGEList(counts=counts(sc), samples=colData(sc))
  if(length(which(table(y$samples$condition) < 2)) > 0){
    return('insufficient cells per sample')}else{
      keep <- filterByExpr(y, group=sc$condition, min.count=filtexprsmin, min.total.count=filtexprstotal)
      y <- y[keep,]
      y <- calcNormFactors(y)
      y$samples$condition <- factor(y$samples$condition, levels=c('PBS', 'Ad-HA/NP', 'Ad-HA/NP+Ad-IL1b', 'pH1N1'))
      par(mfrow=c(2,3))
      for (i in seq_len(ncol(y))) {
        plotMD(y, column=i, main=clust)
      }
      plotMDS(cpm(y, log=TRUE), 
              labels=y$samples$sex, 
              col=c('black', 'blue', 'turquoise', 'red')[as.numeric(factor(y$samples$condition, levels=c('PBS', 'Ad-HA/NP', 'Ad-HA/NP+Ad-IL1b', 'pH1N1')))], main=clust)
      design <- model.matrix(~factor(sex) + factor(condition, levels=c('PBS', 'Ad-HA/NP', 'Ad-HA/NP+Ad-IL1b', 'pH1N1')), y$samples)
      tc <- tryCatch(estimateDisp(y, design), error=function(e) e, warning=function(w) w)
      if(is(tc, 'error') | is(tc, 'warning')){return(paste(tc))}else{
        y <- estimateDisp(y, design)
        plotBCV(y, main=clust)
        fit <- glmQLFit(y, design, robust=TRUE)
        plotQLDisp(fit)
        res <- glmQLFTest(fit, coef=coeffs)
        return(res$table)
      }
    }
}

clusts <- as.character(levels(factor(summed.filt$cluster)))
de.results.any <- vector('list', length(clusts))
names(de.results.any) <- clusts

for(i in 1:length(clusts)){
  de.results.any[[i]] <- pseudoBulkDGEmod(clusts[i], summed.filt, coeffs=3:5)
}


cls <- lapply(de.results.any, class)
if(length(which(cls == 'character'))>0){
  de.results.any <- de.results.any[-which(cls %in% 'character')]
}


for(i in 1:length(de.results.any)){
  de.results.any[[i]]$FDR <- p.adjust(de.results.any[[i]]$PValue, method='fdr')
  de.results.any[[i]]$symbol <- rowData(sce_all_filt)[rownames(de.results.any[[i]]), 'Symbol']
  colnames(de.results.any[[i]])[1:3] <- c('logFC_Ad.HA.NP', 'logFC_Ad.HA.NP.Ad.IL1b', 'logFC_pH1N1')
  write.table(de.results.any[[i]], 
              file=paste0('Results/DE_all_cluster_', names(de.results.any)[i], '.txt'), 
              sep='\t', quote=FALSE)
}
for(i in 1:length(de.results.any)){
  print(head(de.results.any[[i]][de.results.any[[i]]$FDR < 0.05,]))
}


```

There aren't big shifts by condition but there are a handful of genes in most clusters that significantly vary in one or more conditions. 

We would now like to formally test each condition against PBS and generate a list of genes significant in each of these. 

```{r}

de.results.AdHANP <- vector('list', length(clusts))
names(de.results.AdHANP) <- clusts

for(i in 1:length(clusts)){
  de.results.AdHANP[[i]] <- pseudoBulkDGEmod(clusts[i], summed.filt, coeffs=3)
}

cls <- lapply(de.results.AdHANP, class)
if(length(which(cls == 'character'))>0){
  de.results.AdHANP <- de.results.AdHANP[-which(cls %in% 'character')]
}

for(i in 1:length(de.results.AdHANP)){
  de.results.AdHANP[[i]]$FDR <- p.adjust(de.results.AdHANP[[i]]$PValue, method='fdr')
  de.results.AdHANP[[i]]$symbol <- rowData(sce_all_filt)[rownames(de.results.AdHANP[[i]]), 'Symbol']
  write.table(de.results.AdHANP[[i]], 
              file=paste0('Results/DE_AdHANP_cluster_', names(de.results.AdHANP)[i], '.txt'), 
              sep='\t', quote=FALSE)
}
for(i in 1:length(de.results.AdHANP)){
  print(head(de.results.AdHANP[[i]][de.results.AdHANP[[i]]$FDR < 0.05,]))
}




de.results.AdHANPAdIL1b <- vector('list', length(clusts))
names(de.results.AdHANPAdIL1b) <- clusts

for(i in 1:length(clusts)){
  de.results.AdHANPAdIL1b[[i]] <- pseudoBulkDGEmod(clusts[i], summed.filt, coeffs=4)
}

cls <- lapply(de.results.AdHANPAdIL1b, class)
if(length(which(cls == 'character'))>0){
  de.results.AdHANPAdIL1b <- de.results.AdHANPAdIL1b[-which(cls %in% 'character')]
}

for(i in 1:length(de.results.AdHANPAdIL1b)){
  de.results.AdHANPAdIL1b[[i]]$FDR <- p.adjust(de.results.AdHANPAdIL1b[[i]]$PValue, method='fdr')
  de.results.AdHANPAdIL1b[[i]]$symbol <- rowData(sce_all_filt)[rownames(de.results.AdHANPAdIL1b[[i]]), 'Symbol']
  write.table(de.results.AdHANPAdIL1b[[i]], 
              file=paste0('Results/DE_AdHANPAdIL1b_cluster_', names(de.results.AdHANPAdIL1b)[i], '.txt'), 
              sep='\t', quote=FALSE)
}
for(i in 1:length(de.results.AdHANPAdIL1b)){
  print(head(de.results.AdHANPAdIL1b[[i]][de.results.AdHANPAdIL1b[[i]]$FDR < 0.05,]))
}




de.results.pH1N1 <- vector('list', length(clusts))
names(de.results.pH1N1) <- clusts

for(i in 1:length(clusts)){
  de.results.pH1N1[[i]] <- pseudoBulkDGEmod(clusts[i], summed.filt, coeffs=5)
}

cls <- lapply(de.results.pH1N1, class)
if(length(which(cls == 'character'))>0){
  de.results.pH1N1 <- de.results.pH1N1[-which(cls %in% 'character')]
}

for(i in 1:length(de.results.pH1N1)){
  de.results.pH1N1[[i]]$FDR <- p.adjust(de.results.pH1N1[[i]]$PValue, method='fdr')
  de.results.pH1N1[[i]]$symbol <- rowData(sce_all_filt)[rownames(de.results.pH1N1[[i]]), 'Symbol']
  write.table(de.results.pH1N1[[i]], 
              file=paste0('Results/DE_pH1N1_cluster_', names(de.results.pH1N1)[i], '.txt'), 
              sep='\t', quote=FALSE)
}
for(i in 1:length(de.results.pH1N1)){
  print(head(de.results.pH1N1[[i]][de.results.pH1N1[[i]]$FDR < 0.05,]))
}


```

Now we make a few figures for sanity checking and presentation purposes.

Plot IFI6 in all clusters with PBS or pH1N1

```{r}
# example volcano
IL1b.CD8 <- de.results.AdHANPAdIL1b$`6`
cols <- rep('black', times=nrow(IL1b.CD8))
cols[IL1b.CD8$FDR < 0.05] <- 'red'
plot(IL1b.CD8$logFC, -log10(IL1b.CD8$PValue), col=cols)




IFI6 <- 'ENSSSCG00000034570'

plot_resid_gg <- function(y, clust, regress_out, split_by, split_by_levels, col_by, gene, main, ...){
  loggene <- cpm(y, log=TRUE)[gene,y$samples$cluster==clust]
  mod <- lm(loggene~y$samples[y$samples$cluster==clust, regress_out])
  res <- resid(mod)
  tab <- data.frame(res=res, split=y$samples[y$samples$cluster==clust,split_by],
                    col=y$samples[y$samples$cluster==clust,col_by])
  tab$split <- factor(tab$split, levels=split_by_levels)
  ggplot(data=tab, aes(x=split, y=res)) +
    geom_boxplot(na.rm=TRUE) +
    geom_point(aes(col=col), position='jitter') +
    labs(x=NULL, y="expression", col=col_by, title=main) +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, hjust=1), text=element_text(size=21))
}

y <- DGEList(counts=counts(summed.filt), samples=colData(summed.filt))

plot_resid_gg(y, clust='6', regress_out='sex', split_by='condition', split_by_levels=c('PBS', 'Ad-HA/NP', 'Ad-HA/NP+Ad-IL1b', 'pH1N1'), col_by='run', gene=IFI6, main='cluster 6')
plot_resid_gg(y, clust='6', regress_out='run', split_by='condition', split_by_levels=c('PBS', 'Ad-HA/NP', 'Ad-HA/NP+Ad-IL1b', 'pH1N1'), col_by='sex', gene=IFI6, main='cluster 6')
plot_resid_gg(y, clust='6', regress_out='run', split_by='condition', split_by_levels=c('PBS', 'Ad-HA/NP', 'Ad-HA/NP+Ad-IL1b', 'pH1N1'), col_by='run', gene=IFI6, main='cluster 6')

## this is interesting as it has highlighted why we had such different values when including run instead of sex in the model - it looks like one of the pH1N1 samples that was in its own run (other sample was excluded) ended up getting flattened.

for (cluster in unique(y$samples$cluster)){
  plot <- plot_resid_gg(y, clust=cluster, regress_out='sex', split_by='condition', split_by_levels=c('PBS', 'Ad-HA/NP', 'Ad-HA/NP+Ad-IL1b', 'pH1N1'), col_by='sex', gene=IFI6, main=cluster)
  print(plot)
}
# IFI6 appears to be universally up-regulated in infected pigs, though clusters 16 and 21 also show upregulation in IL1b vaccinated pigs. 

pdf('Results/IFI6_pH1N1_egs.pdf', height=10, width=20)
p4 <- plot_resid_gg(y, clust='6', regress_out='sex', split_by='condition', split_by_levels=c('PBS', 'Ad-HA/NP', 'Ad-HA/NP+Ad-IL1b', 'pH1N1'), col_by='sex', gene=IFI6, main='CD8+ T Cells, cluster 6')
p2 <- plot_resid_gg(y, clust='2', regress_out='sex', split_by='condition', split_by_levels=c('PBS', 'Ad-HA/NP', 'Ad-HA/NP+Ad-IL1b', 'pH1N1'), col_by='sex', gene=IFI6, main='Plasma Cells, cluster 2')
p5 <- plot_resid_gg(y, clust='1', regress_out='sex', split_by='condition', split_by_levels=c('PBS', 'Ad-HA/NP', 'Ad-HA/NP+Ad-IL1b', 'pH1N1'), col_by='sex', gene=IFI6, main='Macrophages, cluster 1')
p1 <- plot_resid_gg(y, clust='12', regress_out='sex', split_by='condition', split_by_levels=c('PBS', 'Ad-HA/NP', 'Ad-HA/NP+Ad-IL1b', 'pH1N1'), col_by='sex', gene=IFI6, main='B Cells, cluster 12')
p3 <- plot_resid_gg(y, clust='7', regress_out='sex', split_by='condition', split_by_levels=c('PBS', 'Ad-HA/NP', 'Ad-HA/NP+Ad-IL1b', 'pH1N1'), col_by='sex', gene=IFI6, main='CD4+ T Cells, cluster 7')
p6 <- plot_resid_gg(y, clust='13', regress_out='sex', split_by='condition', split_by_levels=c('PBS', 'Ad-HA/NP', 'Ad-HA/NP+Ad-IL1b', 'pH1N1'), col_by='sex', gene=IFI6, main='cDCs, cluster 13')
grid.arrange(p1, p2, p3, p4, p5, p6, ncol=3)
dev.off()

```




# Wrapping up

Save everything

```{r}
# Update shiny app to filtered data for easier visualization and identification
sce_all_filt$log2_sum_counts <- log2(sce_all_filt$sum_counts)
sce_all_filt$log2_genes_detected <- log2(sce_all_filt$genes_detected)
saveRDS(sce_all_filt, file='shiny_app/data.rds')

sessionInfo()
save(sce_all_filt, file='Data/sce_all_filt.RData')
save(list=ls(), file='ongoing_full_diff.RData') 
```



