# PirbrightBAL

This repository contains analysis markdown files and scripts for analysing porcine BAL scRNAseq data in collaboration with the Pirbright Institute. Now published in PLOS Pathogens titled; 'Single-cell analysis reveals lasting immunological consequences of influenza infection and respiratory immunisation in the pig lung'.

Original unprocessed data can be downloaded from GEO(GSE249866).

To run the first script 'processing all.Rmd', download and place the 'feature_bc_matrix.h5' files from GEO into their appropriate sample folders. Remove the 'sample#_' suffix from each file. Use --- to generate 

Directory should look like this for each sample folder:
-sample1 #folder
  --molecule_info.h5
  --raw_feature_bc_matrix.h5 #available on GEO
  --filtered_feature_bc_matrix.h5 #available on GEO
  --raw_feature_bc_matrix #folder
    ---barcodes.tsv.gz
    ---features.tsv.gz
    ---matrix.mtx.gz
  --filtered_feature_bc_matrix #folder
    ---barcodes.tsv.gz
    ---features.tsv.gz
    ---matrix.mtx.gz

Note that the 'sample1_' suffix in 'sample1_raw_feature_bc_matrix.h5' and 'sample1_filtered_feature_bc_matrix.h5' have been removed.

Scripts in this directory are run in the following order:

processing all.Rmd #Reads in cellranger output (see above) and performs initial QC
clustering all.Rmd #
differential all.Rmd #First look at differential expression, utilises the pseudobulk analysis shown in figure 

The following scripts can be run in any order, provided the scripts above have been run. 

Validation_vs_PBMC.Rmd #
diff-co_NEBULA.Rmd #  differential co-expression analysis with NEBULA. 
Treg_pseudobulk_volcanoes.Rmd # Short script to generate volcano plots of differentially expressed genes from Tregs derived from the pseudobulk analysis in 'differential_all.Rmd'. This script was used for responding  to reviewers comments and doesn't generate data for the related paper.

'shiny_app/data.rds' to use for data exploration in the shiny app ('shiny_app/app.R') is generated by clustering all.Rmd
