---
title: "Validation_vs_PBMC"
author: "Andrew_Muir"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Intro

Continuing from 'differential all.Rmd'.

This script performs a familiarity analysis between the clusters and cell types identified so far in the PirbrightBAL analysis against the reference transcriptome published by Herrera-Uribe et al. 2021 (https://doi.org/10.3389%2Ffgene.2021.689406). 

This includes a mix of cell-type mapping, cluster mapping and cell to cell mapping. 

**Note: This script is compute heavy (due to multiple uses of mnnCorrect) so it is strongly recommended to run on a compute-cluster/workstation with more than 32 GB of memory and at least 16cores.**

# Load packages and data
pigBALsce is from 'differential all.Rmd'.
PBMC's are from:
https://data.nal.usda.gov/dataset/data-reference-transcriptomics-porcine-peripheral-immune-cells-created-through-bulk-and-single-cell-rna-sequencing/resource/7c096891-2807-4e4e-8f5f-0c02c6f434c4

Packages
```{r load-packages}
library(SingleCellExperiment)
library(Seurat)
library(SeuratDisk)
library(scmap)
library(scater)
library(scran)
library(batchelor)
```
Data
```{r load-data}
load('Data/sce_all_filt.RData')
pigBALsce <- sce_all_filt
rm(sce_all_filt)

pigPBMC <- LoadH5Seurat('Herrera-Uribe/PBMC7_AllCells.h5seurat')
```
OR load workspace from previous session. 
```{r load-workspace}
#load('Data/CellTypeSave.RData') #post pre-processing and cell type validation
#load('Data/mnnCorrectSave.RData') #Post mnnCorrection joint UMAP
#load('Data/sankey_Save.RData') #post scmap cluster mapping
#load('Data/Validation_vs_PBMC_Save.RData') #Everything
```

# Basic comparison of datasets

```{r dataset-metrics}
#Check size and depth
dim(pigBALsce)
dim(pigPBMC)

#Check identified cell types
unique(pigBALsce$named_cluster)
unique(pigPBMC$celltypes)

#Check number of clusters
unique(pigBALsce$cluster)
unique(pigPBMC$seurat_clusters)

# Check scale to see if additional normalisation is required
range(counts(pigBALsce))
range(GetAssayData(object = pigPBMC, slot = "counts"))
mean(counts(pigBALsce))
mean(GetAssayData(object = pigPBMC, slot = "counts"))

#p1 <- hist(counts(pigBALsce)[counts(pigBALsce) != 0])
#p2 <- hist(GetAssayData(object = pigPBMC, slot = "counts")[GetAssayData(object = pigPBMC, slot = "counts") != 0])
#plot( p1, col=rgb(0,0,1,1/4), xlim=c(0,100))  
#plot( p2, col=rgb(1,0,0,1/4), xlim=c(0,100), add=T)

```
BAL dataset has 3000 more genes and twice as many cells as PBMC. 

Some good overlap in the important cell types, though the BAL has several cell subtypes that are combined in the PBMC dataset (e.g. mitotic T cells and plasma cells). Obviously this is partially due to the different tissues. We'll deal with this later. 

19 vs 36 clusters, though BAL has had doublets/low-quality clusters removed post clustering. Again deal with synchronization later.

# Check alignment and gene annotation
Both datasets are aligned to the same pig genome from ENSEMBL, albeit different release versions of ENSEMBL (V97 vs V107) and used different alignment functions. This shouldn't be an issue, though we'll do a quick check just in case. 

pigBAL
(cellranger-7.0.0) using default parameters and Sus scrofa genome (genome assembly 11.1, Ensembl release 107).
Babraham pigs

pigPBMC
Sscrofa 11.1 (V97; http://ftp.ensembl.org/pub/release-97/fasta/sus_scrofa/dna/) and annotation GTF (v11.1.97; http://ftp.ensembl.org/pub/release-97/gtf/sus_scrofa/)
custom Perl scripts
Large White and Landrace pigs

Incidentally, both Landrace and Large White pigs were included in the nine breeds used to generate genome assembly 11.1, Pirbright pigs were not. 

```{r check-initial-shared-genes}
PBMC_genes <- rownames(pigPBMC@assays$RNA@counts)

# Try ENSEMBL IDs 
sum(rownames(pigBALsce) %in% PBMC_genes)/nrow(pigBALsce)
sum(PBMC_genes %in% rownames(pigBALsce))/nrow(pigPBMC)

# Try HGNC symbols
sum(rowData(pigBALsce)$hgnc_symbol %in% PBMC_genes)/nrow(pigBALsce)
sum(PBMC_genes %in% rowData(pigBALsce)$hgnc_symbol)/nrow(pigPBMC)

# BAL Symbol column from Cellranger
sum(rowData(pigBALsce)$Symbol %in% PBMC_genes)/nrow(pigBALsce)
sum(PBMC_genes %in% rowData(pigBALsce)$Symbol)/nrow(pigPBMC)

```
ENSEMBL IDs don't map at all, 10% and 12%. 

56% and 68% shared genes using HGNC symbols, not great but not a disaster either. 

PBMC gene annotations are a mix of ENSEMBL IDs and HGNC symbols (in a single slot). Hence Cellranger's 'Symbol' output which is a similar mix works best at 67% and 81% respectively. 

Hence we'll use the Cellranger 'Symbol' output to identify shared genes.    

# Quick check of HVGs
```{r HVG-check}
rownames(pigBALsce) <- rowData(pigBALsce)$Symbol
top_HVGs_BAL <- getTopHVGs(pigBALsce, n=1000)
pigPBMC <- FindVariableFeatures(pigPBMC, nfeatures=1000)
top_HVGs_PBMC <- VariableFeatures(pigPBMC)

sum(top_HVGs_BAL %in% top_HVGs_PBMC)/length(top_HVGs_BAL)
sum(top_HVGs_PBMC %in% top_HVGs_BAL)/length(top_HVGs_PBMC)
```
Only 20% shared HGVs, not great but we'll see if it poses an issue later. 

# Standardisation for scmap
The first analysis we will be running is scmap, which requires that both datasets are sce, that data is stored in specific named slots and that some slots are cleared. 

```{r create-scmap-compatible-sce}
# Remove forbidden rowData names (for generation of CompressedGRangesList)
names(rowData(pigBALsce))[which(names(rowData(pigBALsce))=="start")]="gene_start"
names(rowData(pigBALsce))[which(names(rowData(pigBALsce))=="end")]="gene_end"

# Add slots to BAL
rowData(pigBALsce)$feature_symbol <- rowData(pigBALsce)$Symbol
colData(pigBALsce)$cell_type1 <- pigBALsce$cluster

# Create PBMC sce from seurat object
pigPBMCsce <- as.SingleCellExperiment(pigPBMC)

# Add PBMC slots 
rowData(pigPBMCsce)$feature_symbol <- rownames(pigPBMC@assays$RNA@counts)
colData(pigPBMCsce)$cell_type1 <- as.character(pigPBMC$seurat_clusters)

# Remove genes with zero expression
pigBALsce = pigBALsce[rowSums(counts(pigBALsce)) != 0, ]
pigPBMCsce = pigPBMCsce[rowSums(counts(pigPBMCsce)) != 0, ]

# Restrict both datasets to shared genes only
length(rownames(pigBALsce))
pigBALsce <- pigBALsce[rowData(pigBALsce)$feature_symbol %in% rowData(pigPBMCsce)$feature_symbol]
length(rownames(pigBALsce))

length(rownames(pigPBMCsce))
pigPBMCsce <- pigPBMCsce[rowData(pigPBMCsce)$feature_symbol %in% rowData(pigBALsce)$feature_symbol]
length(rownames(pigPBMCsce))

# Create label and colour slots for convenient annotation later
pigPBMCsce$verbose_cluster <- paste0('PBMC, ', pigPBMCsce$celltypes, ', ', pigPBMCsce$seurat_clusters)
pigPBMCsce$EXP_celltype <- paste0('PBMC, ', pigPBMCsce$celltypes)

PBMC_colour_list <- vector("list", length(pigPBMCsce$EXP_celltype))
unique(pigPBMCsce$EXP_celltype)
PBMC_colour_list[grep('Monocytes', pigPBMCsce$EXP_celltype)] <- '#FB9A99' #pink
PBMC_colour_list[grep('cDCs', pigPBMCsce$EXP_celltype)] <- '#6A3D9A' #purple
PBMC_colour_list[grep('pDCs', pigPBMCsce$EXP_celltype)] <- '#6A3D9A' #purple
PBMC_colour_list[grep('ASC', pigPBMCsce$EXP_celltype)] <- '#1F78B4' #dark blue
PBMC_colour_list[grep('B cells', pigPBMCsce$EXP_celltype)] <- '#A6CEE3' #pale blue
PBMC_colour_list[grep('CD4+', pigPBMCsce$EXP_celltype)] <- '#B2DF8A' #light green
PBMC_colour_list[grep('CD8', pigPBMCsce$EXP_celltype)] <- '#33A02C' # green
PBMC_colour_list[grep('NK cells', pigPBMCsce$EXP_celltype)] <- '#FFFF00' #yellow
PBMC_colour_list[grep('CD2', pigPBMCsce$EXP_celltype)] <- '#CAB2D6' #light purple
PBMC_colour_list[sapply(PBMC_colour_list, is.null)] <- '#B15928' #brown
pigPBMCsce$colour <- PBMC_colour_list

# Set rownames to Symbols for annotation convenience
rownames(pigPBMCsce) <- rowData(pigPBMCsce)$feature_symbol
rownames(pigBALsce) <- rowData(pigBALsce)$feature_symbol

# Normalise by cluster
pigBALsce <- computePooledFactors(pigBALsce, clusters=pigBALsce$cluster, assay.type="counts")
pigBALsce <- logNormCounts(pigBALsce)
pigPBMCsce <- computePooledFactors(pigPBMCsce, cluster=pigPBMCsce$seurat_clusters, assay.type="counts")
pigPBMCsce <- logNormCounts(pigPBMCsce)

# Select gene features for scmap and mnnCorrect, batch corrected
BAL_score_genes <- scoreMarkers(pigBALsce, groups=colData(pigBALsce)$cell_type1, block=colData(pigBALsce)$run)
 
BAL_feature_genes <- vector("list", length(BAL_score_genes))
for(i in 1:length(BAL_score_genes)){
  BAL_feature_genes[[i]] <- BAL_score_genes[[i]][order(BAL_score_genes[[i]]$mean.logFC.cohen, decreasing=TRUE),]
}
names(BAL_feature_genes) <- names(BAL_score_genes)

PBMC_score_genes <- scoreMarkers(pigPBMCsce, groups=colData(pigPBMCsce)$cell_type1, block=colData(pigPBMCsce)$SampleID)

PBMC_feature_genes <- vector("list", length(PBMC_score_genes))
for(i in 1:length(PBMC_score_genes)){
  PBMC_feature_genes[[i]] <- PBMC_score_genes[[i]][order(PBMC_score_genes[[i]]$mean.logFC.cohen, decreasing=TRUE),]
}
names(PBMC_feature_genes) <- names(PBMC_score_genes)

cutoff <- 20
long_cutoff <- 50

BAL_gene_shortlist <- NULL
BAL_gene_longlist <- NULL
BAL_cluster_shortlist <- NULL
for (n in 1:length(BAL_feature_genes)){
  top_genes <- rownames(BAL_feature_genes[[n]])[1:cutoff]
  more_top_genes <- rownames(BAL_feature_genes[[n]])[1:long_cutoff]
  BAL_gene_shortlist <- c(BAL_gene_shortlist, top_genes)
  BAL_gene_longlist <- c(BAL_gene_longlist, more_top_genes)
  cluster_list <- rep(names(BAL_feature_genes)[n], length(top_genes))
  BAL_cluster_shortlist <- c(BAL_cluster_shortlist, cluster_list)
}
BAL_feature_genes_df <- data.frame(matrix(NA, nrow = cutoff*length(BAL_feature_genes), ncol = 2))
colnames(BAL_feature_genes_df) <- c('cluster', 'top_genes')
BAL_feature_genes_df$cluster <- BAL_cluster_shortlist
BAL_feature_genes_df$top_genes <- BAL_gene_shortlist
BAL_gene_shortlist <- unique(BAL_gene_shortlist)
BAL_gene_longlist <- unique(BAL_gene_longlist)

PBMC_gene_shortlist <- NULL
PBMC_gene_longlist <- NULL
PBMC_cluster_shortlist <- NULL
for (n in 1:length(PBMC_feature_genes)){
  top_genes <- rownames(PBMC_feature_genes[[n]])[1:cutoff]
  more_top_genes <- rownames(PBMC_feature_genes[[n]])[1:long_cutoff]
  PBMC_gene_shortlist <- c(PBMC_gene_shortlist, top_genes)
  PBMC_gene_longlist <- c(PBMC_gene_longlist, more_top_genes)
  cluster_list <- rep(names(PBMC_feature_genes)[n], length(top_genes))
  PBMC_cluster_shortlist <- c(PBMC_cluster_shortlist, cluster_list)  
}
PBMC_feature_genes_df <- data.frame(matrix(NA, nrow = cutoff*length(PBMC_feature_genes), ncol = 2))
colnames(PBMC_feature_genes_df) <- c('cluster', 'top_genes')
PBMC_feature_genes_df$cluster <- PBMC_cluster_shortlist
PBMC_feature_genes_df$top_genes <- PBMC_gene_shortlist
PBMC_gene_shortlist <- unique(PBMC_gene_shortlist)
PBMC_gene_longlist <- unique(PBMC_gene_longlist)
```

```{r visualise-new-sce-data, eval=FALSE}
#Sanity check, visualise the converted data, attempt to recreate published visuals
set.seed(123)
pigPBMCsce <- fixedPCA(pigPBMCsce, subset.row=NULL)
pigPBMCsce <- runUMAP(pigPBMCsce, dimred='PCA', n_dimred = 1:10, n_neighbors=20)
#UMAPs
plotUMAP(pigPBMCsce, colour_by="seurat_clusters", point_size=0.3) + guides(colour = guide_legend(override.aes = list(size=8)))+ theme(text = element_text(size = 20))
plotUMAP(pigPBMCsce, colour_by="celltypes", point_size=0.3) + guides(colour = guide_legend(override.aes = list(size=8)))+ theme(text = element_text(size = 20))
# Highlight T cells, check if correct annotation
plotUMAP(pigPBMCsce, colour_by="CD3E", point_size=0.3) + guides(colour = guide_legend(override.aes = list(size=8)))+ theme(text = element_text(size = 20))
# Highlight monocytes, check if correct annotation
plotUMAP(pigPBMCsce, colour_by="CD14", point_size=0.3) + guides(colour = guide_legend(override.aes = list(size=8)))+ theme(text = element_text(size = 20))
# Highlight B cells, check if correct annotation
plotUMAP(pigPBMCsce, colour_by="CD19", point_size=0.3) + guides(colour = guide_legend(override.aes = list(size=8)))+ theme(text = element_text(size = 20))
# Highlight SampleID to see if data is already batch corrected
plotUMAP(pigPBMCsce, colour_by="SampleID", point_size=0.3) + guides(colour = guide_legend(override.aes = list(size=8)))+ theme(text = element_text(size = 20))
```
Plots aren't the cleanest in terms of cluster localisation (compared to published), however CD3E, CD14 and CD19 appear to localise to the correct cell types, so the conversion hasn't broken anything and we're okay to move on. 

```{r cell-type-save}
save(list=ls(), file='Data/CellTypeSave.RData') 
```

# mnnCorrect
Genes in both datasets need to be in identical order. 
```{r standardise-for-mnnCorrect}
pigPBMCsce <- pigPBMCsce[match(rowData(pigBALsce)$feature_symbol, rowData(pigPBMCsce)$feature_symbol),]
head(rowData(pigBALsce)$feature_symbol, n=10)
head(rowData(pigPBMCsce)$feature_symbol, n=10)
tail(rownames(pigBALsce), n=10)
tail(rownames(pigPBMCsce), n=10)
identical(rownames(pigPBMCsce), rownames(pigBALsce))
identical(rowData(pigBALsce)$feature_symbol, rowData(pigPBMCsce)$feature_symbol)
```
Restricting genes fed to mnnCorrect to improve merging and speed up computation. 
```{r trim-genes-mnncorrect}
Union_feature_genes <- paste0(c(BAL_gene_shortlist, PBMC_gene_shortlist))
Union_feature_genes <- unique(Union_feature_genes)

BAL_Union_sce <- pigBALsce[rownames(pigBALsce) %in% Union_feature_genes, ]
PBMC_Union_sce <- pigPBMCsce[rownames(pigPBMCsce) %in% Union_feature_genes, ]
identical(rownames(BAL_Union_sce), rownames(PBMC_Union_sce))
```
split into batches, to correct for intra-dataset differences prior to between dataset differences

Quoting PBMC paper, on batches
"Experiment B (Exp B) for NanoString and scRNA-seq from three ∼12-month-old pigs (B1, B2, B3); Experiment C (ExpC) for scRNA-seq from three ∼12-month-old pigs (C1, C2, C3); Experiment D (ExpD) for scRNA-seq from two ∼7-week-old pigs (D1, D2)."

Internal	Publication
CT230OCT = B1
CT21NOV	= B2
A =	C1
B =	C2
C =	C3
PBMC1 =	D1
PBMC2 =	D2
```{r split-batches}
#BAL
BAL_Union_sce_run1 <- BAL_Union_sce[, BAL_Union_sce$run == "run1"]
BAL_Union_sce_run2 <- BAL_Union_sce[, BAL_Union_sce$run == "run2"]
BAL_Union_sce_run3 <- BAL_Union_sce[, BAL_Union_sce$run == "run3"]
BAL_Union_sce_run4 <- BAL_Union_sce[, BAL_Union_sce$run == "run4"]
BAL_Union_sce_run5 <- BAL_Union_sce[, BAL_Union_sce$run == "run5"]
BAL_Union_sce_run6 <- BAL_Union_sce[, BAL_Union_sce$run == "run6"]

#PBMC
PBMC_Union_sce_B <- PBMC_Union_sce[, PBMC_Union_sce$SampleID %in% c("CT230OCT", "CT21NOV")]
PBMC_Union_sce_C <- PBMC_Union_sce[, PBMC_Union_sce$SampleID %in% c("A", "B", "C")]
PBMC_Union_sce_D <- PBMC_Union_sce[, PBMC_Union_sce$SampleID %in% c("PBMC1", "PBMC2")]
```

Run mnnCorrect
**Warning: this step is compute heavy.**
```{r run-mnnCorrect}
mnnCorrected <- mnnCorrect(logcounts(BAL_Union_sce_run1), logcounts(BAL_Union_sce_run2), logcounts(BAL_Union_sce_run3), logcounts(BAL_Union_sce_run4), logcounts(BAL_Union_sce_run5), logcounts(BAL_Union_sce_run6), logcounts(PBMC_Union_sce_B), logcounts(PBMC_Union_sce_C), logcounts(PBMC_Union_sce_D), k=20, merge.order=list(list(1,2,3,4,5,6), list(7,8,9)))
save(mnnCorrected, file='Data/mnnCorrected_feature20_k20.RData')
#load('Data/mnnCorrected_feature20_k20.RData')
mnnCorrected
```

Return mnn pairs, evaluate if sufficent number (else adjust k)
```{r check-mnn-pairs}
metadata(mnnCorrected)$merge.info

colData(mnnCorrected)
```

# Joint analysis

```{r joint-TSNE-UMAP}
# Add labels and colours for visualisation
verbose_labels <- c(pigBALsce$verbose_cluster, pigPBMCsce$verbose_cluster)
mnnCorrected$verbose_cluster <- verbose_labels 
EXP_celltype_labels <- c(pigBALsce$EXP_celltype, pigPBMCsce$EXP_celltype)
mnnCorrected$EXP_celltype <- EXP_celltype_labels
colour_labels <- c(pigBALsce$colour, pigPBMCsce$colour)
mnnCorrected$colour <- colour_labels
mnnCorrected$batch[mnnCorrected$batch %in% c(1,2,3,4,5,6)] <- "BAL"
mnnCorrected$batch[mnnCorrected$batch %in% c(7,8,9)] <- "PBMC"
colData(mnnCorrected)

legend_colours <- colour_labels
names(legend_colours) <- mnnCorrected$EXP_celltype
legend_colours <- legend_colours[!duplicated(names(legend_colours))]
legend_colours <- legend_colours[c("BAL, Macrophages", "BAL, Monocytes", "PBMC, Monocytes", "BAL, cDCs", "PBMC, cDCs", "PBMC, pDCs", "BAL, Plasma Cells", "PBMC, ASC", "BAL, B Cells", "PBMC, B cells", "BAL, CD4 T Cells", "PBMC, CD4+ ab T cells", "BAL, Tregs", "BAL, Mitotic CD4 T Cells", "BAL, Mitotic CD8 T Cells", "BAL, CD8 T Cells", "PBMC, CD8ab+ ab T cells", "PBMC, CD8a+ ab T/NK cells", "PBMC, NK cells", "PBMC, CD2+ GD T cells", "PBMC, CD2- GD T cells", "BAL, T-cell-like", "BAL, Unknown", "PBMC, Unknown", "PBMC, Erythrocytes")]

df_legend_colours <- data.frame(unlist(names(legend_colours)), unlist(legend_colours))
names(df_legend_colours) <- c("Cell_Type", "Colour")
df_legend_colours$dummy_data <- seq(1, length(df_legend_colours[,1]))

# T cell subset
mnnCorrected_T_Cells <- mnnCorrected[, mnnCorrected$EXP_celltype %in% c("BAL, CD8 T Cells", "BAL, CD4 T Cells", "BAL, Tregs", "BAL, Mitotic CD8 T Cells", "BAL, Mitotic CD4 T Cells", "PBMC, CD4+ ab T cells", "PBMC, CD2- GD T cells", "PBMC, CD8a+ ab T/NK cells", "PBMC, CD2+ GD T cells", "PBMC, CD8ab+ ab T cells")]

#Calculate UMAPs from PCA
set.seed(123)
mnnCorrected <- fixedPCA(mnnCorrected, subset.row=NULL, rank = 20, assay.type='corrected')
mnnCorrected <- runUMAP(mnnCorrected, dimred='PCA', n_dimred = 1:10, n_neighbors=20)

mnnCorrected_T_Cells <- fixedPCA(mnnCorrected_T_Cells, subset.row=NULL, rank = 20, assay.type='corrected')
mnnCorrected_T_Cells <- runUMAP(mnnCorrected_T_Cells, dimred='PCA', n_dimred = 1:10, n_neighbors=20)

library(gridExtra)

png(filename='Results/Combined_mnnCorrectedUMAP.png', height=2160, width=4320)
plot0 <- ggplot(data=df_legend_colours, aes(x=dummy_data)) +
  geom_point(aes(y=dummy_data, color=Cell_Type), size=10) +
  labs(x=NULL, y=NULL, color="A") +
  scale_color_manual(values = legend_colours, breaks=names(legend_colours)) +
  guides(colour = guide_legend(ncol = 1)) + 
  theme(legend.position=c(-0.35,0), legend.direction="vertical", aspect.ratio=0, text = element_text(size = 40),
        axis.text.x=element_blank(), axis.ticks.x=element_blank(),
        axis.text.y=element_blank(), axis.ticks.y=element_blank())
plot1 <- plotUMAP(mnnCorrected, point_size=0) +
  aes(colour = mnnCorrected$colour) +
  geom_point(shape='.', size=0.3) +
  geom_text(x=-12, y=10, size=30, label="A", color="black") +
  theme(aspect.ratio = 1, plot.margin = unit(c(0,0,0,-35), "cm"), text = element_text(size = 40))
plot2 <- plotUMAP(mnnCorrected, colour_by="batch", point_size=0.3) +
  theme(legend.position=c(0.8,0.8), legend.direction="vertical") +
  geom_text(x=-12, y=10, size=30, label="B") +
  guides(colour = guide_legend(override.aes = list(size=10), title="B"))+ theme(text = element_text(size = 40)) +
  theme(aspect.ratio = 1, plot.margin = unit(c(0,0,0,-35), "cm"))
plot3 <- plotUMAP(mnnCorrected_T_Cells, colour_by="EXP_celltype", point_size=0.3) +
  theme(legend.position=c(0.5,1.2), legend.direction="vertical") +
  geom_text(x=-12, y=11, size=30, label="C") +
  guides(colour = guide_legend(override.aes = list(size=10), title="C", ncol=1))+ theme(text = element_text(size = 40)) +
  theme(aspect.ratio = 1, plot.margin = unit(c(0,0,0,-110), "cm"))
grid.arrange(plot1, plot0, plot2, plot3, ncol=2)
dev.off()

```

mnncorrect save
```{r mnncorrect-save}
save(list=ls(), file='Data/mnnCorrectSave.RData')
```

scmap on batch corrected data
```{r batch-correct-scmap}
corrected_BAL <- mnnCorrect(BAL_Union_sce, batch=BAL_Union_sce$run, k=20)
corrected_PBMC <- mnnCorrect(PBMC_Union_sce, batch=PBMC_Union_sce$SampleID, k=20)

# Set essential data
assay(corrected_BAL, i='logcounts') <- assay(corrected_BAL, i='corrected')
assay(corrected_PBMC, i='logcounts') <- assay(corrected_PBMC, i='corrected')

rowData(corrected_BAL)$feature_symbol <- rowData(BAL_Union_sce)$feature_symbol
colData(corrected_BAL)$cell_type1 <- BAL_Union_sce$cell_type1
colData(corrected_BAL)$verbose_cluster <- BAL_Union_sce$verbose_cluster
colData(corrected_BAL)$colour <- BAL_Union_sce$colour
colData(corrected_BAL)$cluster <- BAL_Union_sce$cluster

rowData(corrected_PBMC)$feature_symbol <- rowData(PBMC_Union_sce)$feature_symbol
colData(corrected_PBMC)$cell_type1 <- PBMC_Union_sce$cell_type1
colData(corrected_PBMC)$verbose_cluster <- PBMC_Union_sce$verbose_cluster
colData(corrected_PBMC)$colour <- PBMC_Union_sce$colour
colData(corrected_PBMC)$seurat_clusters <- PBMC_Union_sce$seurat_clusters

#select features on BAL
corrected_BAL <- setFeatures(corrected_BAL, BAL_gene_shortlist)

#index on BAL
corrected_BAL <- indexCluster(corrected_BAL)

#select features on PBMC
corrected_PBMC <- setFeatures(corrected_PBMC, PBMC_gene_shortlist)

#index on PBMC
corrected_PBMC <- indexCluster(corrected_PBMC)

#project data
corrected_clusterBAL_to_clusterPBMC <- scmapCluster(
  projection = corrected_BAL,
  index_list = list(corrected_PBMC = metadata(corrected_PBMC)$scmap_cluster_index),
  threshold = 0.7
)
#make data tables (to be plotted)
corrected_BAL_proj_table <- table(colData(corrected_BAL)$cell_type1, corrected_clusterBAL_to_clusterPBMC$scmap_cluster_labs)
# Make percentage table
corrected_BAL_percent_proj_table <- round(prop.table(corrected_BAL_proj_table, margin=1),2)
```

make table and plot
```{r plot-corrected-scmap}
corrected_map_table <- corrected_BAL_proj_table
# re-order
corrected_map_table <- corrected_map_table[order(rownames(corrected_map_table), decreasing = FALSE), ]
corrected_map_table <- corrected_map_table[, order(colnames(corrected_map_table), decreasing = FALSE)]

#rename BAL
corrected_BAL_clusters <- unique(data.frame(BAL_Union_sce$cluster, BAL_Union_sce$named_cluster))
corrected_BAL_clusters <- corrected_BAL_clusters[match(rownames(corrected_map_table), corrected_BAL_clusters$BAL_Union_sce.cluster),]

if(identical(rownames(corrected_map_table), corrected_BAL_clusters$BAL_Union_sce.cluster) == FALSE){
  stop("BAL names not identical")
}
corrected_BAL_names <- paste0(corrected_BAL_clusters$BAL_Union_sce.named_cluster, ', ', corrected_BAL_clusters$BAL_Union_sce.cluster)
rownames(corrected_map_table) <- corrected_BAL_names

#rename PBMC
corrected_PBMC_clusters <- unique(data.frame(as.character(PBMC_Union_sce$seurat_clusters), as.character(PBMC_Union_sce$celltypes)))
corrected_PBMC_clusters <- corrected_PBMC_clusters[match(colnames(corrected_map_table), corrected_PBMC_clusters$as.character.PBMC_Union_sce.seurat_clusters),]
corrected_PBMC_clusters[is.na(corrected_PBMC_clusters)] <- 'unassigned'

if(identical(colnames(corrected_map_table), corrected_PBMC_clusters$as.character.PBMC_Union_sce.seurat_clusters) == FALSE){
  stop("PBMC names not identical")
}
corrected_PBMC_names <- paste0(corrected_PBMC_clusters$as.character.PBMC_Union_sce.celltypes, ', ', corrected_PBMC_clusters$as.character.PBMC_Union_sce.seurat_clusters)
colnames(corrected_map_table) <- corrected_PBMC_names

# re-order, now with new names
corrected_map_table <- corrected_map_table[order(rownames(corrected_map_table), decreasing = FALSE), ]
corrected_map_table <- corrected_map_table[, order(colnames(corrected_map_table), decreasing = FALSE)]
write.csv(corrected_map_table, file='Results/corrected_cluster_projection_counts.csv')

# as percentage
corrected_map_percent_table <- round(prop.table(corrected_map_table, margin=1),2)
write.csv(corrected_map_percent_table, file='Results/corrected_cluster_projection_percent.csv')

# Prep for sankey
corrected_map_dataframe <- data.frame(corrected_map_percent_table)
# Trim flow, remove low values for legibility
corrected_map_dataframe <- corrected_map_dataframe[corrected_map_dataframe$Freq > 0.15, ]

library(networkD3)
library(dplyr)
 
# connection data frame
links <- data.frame(
  source=corrected_map_dataframe$Var1, 
  target=corrected_map_dataframe$Var2, 
  value=corrected_map_dataframe$Freq
  )
 
# node data frame
nodes <- data.frame(
  name=c(as.character(links$source), 
  as.character(links$target)) %>% unique()
)
# Group nodes by cell type
grouplist <- vector("list", length(nodes$name))

grouplist[grep('Plasma Cells', nodes$name)] <- 'a'
grouplist[grep('ASC', nodes$name)] <- 'a'
grouplist[grep('B Cells', nodes$name)] <- 'b'
grouplist[grep('B cells', nodes$name)] <- 'b'
grouplist[grep('CD4', nodes$name)] <- 'c'
grouplist[grep('Tregs', nodes$name)] <- 'd'
grouplist[grep('NK', nodes$name)] <- 'f'
grouplist[grep('CD8', nodes$name)] <- 'e'
grouplist[grep('Mitotic CD4', nodes$name)] <- 'j'
grouplist[grep('Mitotic CD8', nodes$name)] <- 'k'
grouplist[grep('Monocytes', nodes$name)] <- 'g'
grouplist[grep('Macrophages', nodes$name)] <- 'h'
grouplist[grep('DCs', nodes$name)] <- 'i'
grouplist[grep('pDCs', nodes$name)] <- 'i'
grouplist[grep('CD2', nodes$name)] <- 'd'
grouplist[sapply(grouplist, is.null)] <- 'l'

nodes$group <- grouplist

# Group links by cell type
grouplist2 <- vector("list", length(links$source))

grouplist2[grep('Plasma Cells', links$source)] <- 'a'
grouplist2[grep('ASC', links$source)] <- 'a'
grouplist2[grep('B Cells', links$source)] <- 'b'
grouplist2[grep('B cells', links$source)] <- 'b'
grouplist2[grep('CD4', links$source)] <- 'c'
grouplist2[grep('Tregs', links$source)] <- 'd'
grouplist2[grep('NK', links$source)] <- 'f'
grouplist2[grep('CD8', links$source)] <- 'e'
grouplist2[grep('Mitotic CD4', links$source)] <- 'j'
grouplist2[grep('Mitotic CD8', links$source)] <- 'k'
grouplist2[grep('Monocytes', links$source)] <- 'g'
grouplist2[grep('Macrophages', links$source)] <- 'h'
grouplist2[grep('DCs', links$source)] <- 'i'
grouplist2[grep('pDCs', links$source)] <- 'i'
grouplist2[grep('CD2', links$source)] <- 'd'
grouplist2[sapply(grouplist2, is.null)] <- 'l'

links$group <- grouplist2

# Give each group a colour
my_color <- 'd3.scaleOrdinal(d3.schemeCategory10) .domain(["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l"]) .range(["#1F78B4", "#A6CEE3", "#B2DF8A", "#CAB2D6", "#33A02C", "#FFFF00", "#FB9A99", "#E31A1C", "#6A3D9A",  "#FDBF6F", "#FF7F00", "#B15928"])'

# convert names to IDs
links$IDsource <- match(links$source, nodes$name)-1 
links$IDtarget <- match(links$target, nodes$name)-1
 
# Make the Network
corrected_sanPlot <- sankeyNetwork(Links = links, Nodes = nodes,
              Source = "IDsource", Target = "IDtarget",
              Value = "value", NodeID = "name",
              colourScale=my_color, LinkGroup="group", NodeGroup="group",
              fontFamily = "arial", fontSize = 8, nodeWidth = 8,
              sinksRight=FALSE)
corrected_sanPlot

htmltools::save_html(corrected_sanPlot, file = "Results/corrected_cluster_map_sankey.html")
```

```{r save-sankey}
sessionInfo()
save(list=ls(), file='Data/sankey_Save.RData') 
```

# Heatmaps
Make heatmaps of interesting subsets 
```{r heatmaps}
library(gplots)
library(viridis)

# Everything heatmap
BAL_heat_all <- corrected_BAL[rownames(corrected_BAL) %in% PBMC_gene_shortlist, ]
PBMC_heat_all <- corrected_PBMC[rownames(corrected_PBMC) %in% PBMC_gene_shortlist, ]
identical(rownames(PBMC_heat_all), rownames(BAL_heat_all))

BAL_heat_all <- as.matrix(logcounts(aggregateAcrossCells(BAL_heat_all, ids=BAL_heat_all$verbose_cluster, statistics='mean', use.assay.type='logcounts')))
PBMC_heat_all <- as.matrix(logcounts(aggregateAcrossCells(PBMC_heat_all, ids=PBMC_heat_all$verbose_cluster, statistics='mean', use.assay.type='logcounts')))
heat_all_Matrix <- cbind(BAL_heat_all, PBMC_heat_all)
write.csv(heat_all_Matrix, file='Results/heatmap_scmap_all_raw_values.csv')

#Identify universal top and bottom genes to set scales
gene_scale <- apply(heat_all_Matrix, 1, mean)
names(gene_scale) <- rownames(heat_all_Matrix)
gene_scale <- sort(gene_scale)
head(gene_scale, n=5)
tail(gene_scale, n=5)

gene_scale <- apply(heat_all_Matrix, 1, sum)
names(gene_scale) <- rownames(heat_all_Matrix)
gene_scale <- sort(gene_scale)
head(gene_scale, n=5)
tail(gene_scale, n=5)

# i.e. most range between 0 and 0.2, some outliers reaching for 0.3
#breaks = seq(0,max(heat_all_Matrix),length.out=257)
breaks = seq(0,0.2,length.out=257)

xAxis <- colnames(heat_all_Matrix)
xColours <- ifelse(grepl("BAL", xAxis, fixed=TRUE), "blue", "orange")
colour_df <- rbind(unique(as.data.frame(colData(corrected_BAL))[, c("verbose_cluster", "colour")]), unique(as.data.frame(colData(corrected_PBMC))[, c("verbose_cluster", "colour")]))
colour_df <- colour_df[match(xAxis, colour_df$verbose_cluster),]
identical(colour_df$verbose_cluster, xAxis)
Col_Colours <- as.character(colour_df$colour)

png(filename='Results/heatmap_scmap_all.png', height=2160, width=3840)
heatmap.2(heat_all_Matrix, margins=c(30,50), scale="row", col = bluered(100), cexRow=2, cexCol=2, colCol=xColours,  trace="none", key=TRUE, key.xlab="row Z-score", lhei=c(1,10), lwid=c(1,10), ColSideColors=c(Col_Colours))
dev.off()

# B cells
B_Cell_genes <- c("CD19", "PAX5", "EBF1", "MS4A1", "CD79B", "IRF8", "ENSSSCG00000015664", "ZBTB32", "HVCN1", "BLK", "PLD4", "CD79B", "ENSSSCG00000040849", "SMC6", "HLA-DRA", "SLA-DQB1", "CD74", "RPL9", "RPL6", "RPS18", "ENSSSCG00000013889", "RPL7A", "ENSSSCG00000011272", "RPS11", "ACTB", "UBB", "RPS23", "ENSSSCG00000014540", "RPS3A", "RPS15", "RPL11", "RPS20", "TMSB10")

BAL_B_cells <- corrected_BAL[rownames(corrected_BAL) %in% B_Cell_genes, ]
BAL_B_cells <- BAL_B_cells[, BAL_B_cells$cluster %in% c("12", "16")]
PBMC_B_cells <- corrected_PBMC[rownames(corrected_PBMC) %in% B_Cell_genes, ]
PBMC_B_cells <- PBMC_B_cells[, PBMC_B_cells$seurat_clusters %in% c("8", "16")]
identical(rownames(PBMC_B_cells), rownames(BAL_B_cells))

BAL_B_cells <- as.matrix(logcounts(aggregateAcrossCells(BAL_B_cells, ids=BAL_B_cells$verbose_cluster, statistics='mean', use.assay.type='logcounts')))
PBMC_B_cells <- as.matrix(logcounts(aggregateAcrossCells(PBMC_B_cells, ids=PBMC_B_cells$verbose_cluster, statistics='mean', use.assay.type='logcounts')))
B_Cells_Matrix <- cbind(BAL_B_cells, PBMC_B_cells)

png(filename='Results/heatmap_B_Cells_Cross.png', height=2160, width=1080)
heatmap.2(B_Cells_Matrix, Colv=FALSE, margins=c(50,40), scale="none", trace="none", dendrogram="row", breaks=breaks, col=turbo(256), lhei=c(0.1,10), lwid=c(1,10), cexRow=4, cexCol=4, key=FALSE)
dev.off()

png(filename='Results/heatmap_B_cells_with_scale.png', height=2160, width=1080)
heatmap.2(B_Cells_Matrix, Colv=FALSE, margins=c(15,15), scale="none", trace="none", dendrogram="row", breaks=breaks, col=turbo(256), lhei=c(1,9), key=TRUE, keysize=2, density.info='none', key.xlab="mean logcounts")
dev.off()

# Plasma cells
plasma_genes <- c("PRDM1", "IRF4", "JCHAIN", "TNFRSF17", "TNFRSF13B", "MANF", "ATOX1", "SEC11C", "SSR3", "TXNDC5", "CRELD2", "MZBD1", "ENSSSCG00000038719", "HLA-DRA", "SLA-DQB1", "CD74", "SLA-DMB", "CTSZ", "SNX3", "CD79B", "ENSSSCG00000040849", "SMC6")

BAL_plasma <- corrected_BAL[rownames(corrected_BAL) %in% plasma_genes, ]
BAL_plasma <- BAL_plasma[, BAL_plasma$cluster %in% c("2", "3")]
PBMC_plasma <- corrected_PBMC[rownames(corrected_PBMC) %in% plasma_genes, ]
PBMC_plasma <- PBMC_plasma[, PBMC_plasma$seurat_clusters %in% c("29", "32")]
identical(rownames(PBMC_plasma), rownames(BAL_plasma))

BAL_plasma <- as.matrix(logcounts(aggregateAcrossCells(BAL_plasma, ids=BAL_plasma$verbose_cluster, statistics='mean', use.assay.type='logcounts')))
PBMC_plasma <- as.matrix(logcounts(aggregateAcrossCells(PBMC_plasma, ids=PBMC_plasma$verbose_cluster, statistics='mean', use.assay.type='logcounts')))
plasma_Matrix <- cbind(BAL_plasma, PBMC_plasma)

png(filename='Results/heatmap_plasma_Cross.png', height=2160, width=1080)
heatmap.2(plasma_Matrix, margins=c(50,40), scale="none", trace="none", dendrogram="row", breaks=breaks, col=turbo(256), lhei=c(0.1,10), lwid=c(1,10), cexRow=4, cexCol=4, key=FALSE)
dev.off()

# Mitotic T cells
mitotic_genes <- c("CD28", "CD5", "CD3D", "CD3E", "CD4", "CD8A", "CD8B", "FOXP3", "E2F1", "E2F2", "KLRK1", "KLRB1", "KIT", "NCR1", "TBX21", "EOMES", "GATA3", "IL1RL1", "IL7R", "ID3", "CD5", "EVL", "CD247", "RHEX", "TRAT1", "NKG7", "GZMB", "CD2", "ENSSSCG00000009327", "SAT1", "TK1", "SMC2", "TYMS", "SPC24", "BIRC5", "PCNA", "SMC4", "PCLAF", "TUBA1B", "H2AFZ", "STMN1", "LAT", "SH2D1A", "DAPL1")

BAL_mitotic <- corrected_BAL[rownames(corrected_BAL) %in% mitotic_genes, ]
BAL_mitotic <- BAL_mitotic[, BAL_mitotic$cluster %in% c("18a", "18b")]
PBMC_mitotic <- corrected_PBMC[rownames(corrected_PBMC) %in% mitotic_genes, ]
PBMC_mitotic <- PBMC_mitotic[, PBMC_mitotic$seurat_clusters %in% c("0", "14", "28")]
identical(rownames(PBMC_mitotic), rownames(BAL_mitotic))

BAL_mitotic <- as.matrix(logcounts(aggregateAcrossCells(BAL_mitotic, ids=BAL_mitotic$verbose_cluster, statistics='mean', use.assay.type='logcounts')))
PBMC_mitotic <- as.matrix(logcounts(aggregateAcrossCells(PBMC_mitotic, ids=PBMC_mitotic$verbose_cluster, statistics='mean', use.assay.type='logcounts')))
mitotic_matrix <- cbind(BAL_mitotic, PBMC_mitotic)

png(filename='Results/heatmap_mitotic_T_cells_Cross.png', height=2160, width=1080)
heatmap.2(mitotic_matrix, margins=c(50,40), scale="none", trace="none", dendrogram="row", breaks=breaks, col=turbo(256), lhei=c(0.1,10), lwid=c(1,10), cexRow=4, cexCol=4, key=FALSE)
dev.off()

# Conventional T cells
T_cell_genes <- c("CD28", "CD5", "CD3D", "CD3E", "CD4", "CD8A", "CD8B", "FOXP3", "E2F1", "E2F2", "KLRK1", "KLRB1", "KIT", "NCR1", "TBX21", "EOMES", "GATA3", "IL1RL1", "IL7R", "ID3", "CD5", "EVL", "CD247", "RHEX", "TRAT1", "NKG7", "GZMB", "CD2", "ENSSSCG00000009327", "SAT1", "LAT", "SH2D1A", "DAPL1")

BAL_T_cell <- corrected_BAL[rownames(corrected_BAL) %in% T_cell_genes, ]
BAL_T_cell <- BAL_T_cell[, BAL_T_cell$cluster %in% c("6", "7", "10")]
PBMC_T_cell <- corrected_PBMC[rownames(corrected_PBMC) %in% T_cell_genes, ]
PBMC_T_cell <- PBMC_T_cell[, PBMC_T_cell$seurat_clusters %in% c("4", "21", "28")]
identical(rownames(PBMC_T_cell), rownames(BAL_T_cell))

BAL_T_cell <- as.matrix(logcounts(aggregateAcrossCells(BAL_T_cell, ids=BAL_T_cell$verbose_cluster, statistics='mean', use.assay.type='logcounts')))
PBMC_T_cell <- as.matrix(logcounts(aggregateAcrossCells(PBMC_T_cell, ids=PBMC_T_cell$verbose_cluster, statistics='mean', use.assay.type='logcounts')))
T_Cell_Matrix <- cbind(BAL_T_cell, PBMC_T_cell)

png(filename='Results/heatmap_conventional_T_cells_Cross.png', height=2160, width=1080)
heatmap.2(T_Cell_Matrix, margins=c(50,40), scale="none", trace="none", dendrogram="row", breaks=breaks, col=turbo(256), lhei=c(0.1,10), lwid=c(1,10), cexRow=4, cexCol=4, key=FALSE)
dev.off()

# Monocytic cells
mono_genes <- c("CD163", "FLT3", "CD14", "CD4", "CADM1", "CSF1R", "FCGR1A", "BCL6", "CCL17", "IRF8", "ADAM8", "ANXA5", "CEBPB", "CAPG", "S100A8", "YBX3", "PCNA", "SMC4", "PCLAF", "VIM", "SRGN", "CD63", "TYROBP", "ANXA1", "CTSS", "LYZ", "FOLR1", "IFI6", "CST3")

BAL_mono <- corrected_BAL[rownames(corrected_BAL) %in% mono_genes, ]
BAL_mono <- BAL_mono[, BAL_mono$cluster %in% c("1", "4", "8", "13", "15")]
PBMC_mono <- corrected_PBMC[rownames(corrected_PBMC) %in% mono_genes, ]
PBMC_mono <- PBMC_mono[, PBMC_mono$seurat_clusters %in% c("20", "27", "30", "32")]
identical(rownames(PBMC_mono), rownames(BAL_mono))

BAL_mono <- as.matrix(logcounts(aggregateAcrossCells(BAL_mono, ids=BAL_mono$verbose_cluster, statistics='mean', use.assay.type='logcounts')))
PBMC_mono <- as.matrix(logcounts(aggregateAcrossCells(PBMC_mono, ids=PBMC_mono$verbose_cluster, statistics='mean', use.assay.type='logcounts')))
mono_Matrix <- cbind(BAL_mono, PBMC_mono)

png(filename='Results/heatmap_monocytic_Cross.png', height=2160, width=1080)
heatmap.2(mono_Matrix, margins=c(50,40), scale="none", trace="none", dendrogram="row", breaks=breaks, col=turbo(256), lhei=c(0.1,10), lwid=c(1,10), cexRow=4, cexCol=4, key=FALSE)
dev.off()

# Cluster 20 T cells
BAL_C20 <- corrected_BAL[rownames(corrected_BAL) %in% T_cell_genes, ]
BAL_C20 <- BAL_C20[, BAL_C20$cluster %in% c("20a", "20b")]
PBMC_C20 <- corrected_PBMC[rownames(corrected_PBMC) %in% T_cell_genes, ]
PBMC_C20 <- PBMC_C20[, PBMC_C20$seurat_clusters %in% c("0", "3", "28")]
identical(rownames(PBMC_C20), rownames(BAL_C20))

BAL_C20 <- as.matrix(logcounts(aggregateAcrossCells(BAL_C20, ids=BAL_C20$verbose_cluster, statistics='mean', use.assay.type='logcounts')))
PBMC_C20 <- as.matrix(logcounts(aggregateAcrossCells(PBMC_C20, ids=PBMC_C20$verbose_cluster, statistics='mean', use.assay.type='logcounts')))
C20_Matrix <- cbind(BAL_C20, PBMC_C20)

png(filename='Results/heatmap_cluster20_T_cells_Cross.png', height=2160, width=1080)
heatmap.2(C20_Matrix, margins=c(50,40), scale="none", trace="none", dendrogram="row", breaks=breaks, col=turbo(256), lhei=c(0.1,10), lwid=c(1,10), cexRow=4, cexCol=4, key=FALSE)
dev.off()

```
# Gene expression bubbleplots (dotplots)
```{r bubbleplots, eval=FALSE}
library(gridExtra)
theme_set(theme_minimal(base_size = 14))
# B cells
BAL_B_cells <- corrected_BAL[rownames(corrected_BAL) %in% B_Cell_genes, ]
BAL_B_cells <- BAL_B_cells[, BAL_B_cells$cluster %in% c("12", "16")]
PBMC_B_cells <- corrected_PBMC[rownames(corrected_PBMC) %in% B_Cell_genes, ]
PBMC_B_cells <- PBMC_B_cells[, PBMC_B_cells$seurat_clusters %in% c("8", "16")]
rowData(BAL_B_cells)$scmap_features <- NULL
rowData(PBMC_B_cells)$scmap_features <- NULL
colData(BAL_B_cells)$cluster <- NULL
colData(PBMC_B_cells)$seurat_clusters <- NULL
Combined_B_Cells <- cbind(BAL_B_cells, PBMC_B_cells)

p1 <- plotDots(Combined_B_Cells, features=rownames(Combined_B_Cells), group="verbose_cluster") +
  theme(aspect.ratio = 4, axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.title.x=element_blank(), axis.title.y=element_blank())

# T cells 

BAL_T_cells <- corrected_BAL[rownames(corrected_BAL) %in% T_cell_genes, ]
BAL_T_cells <- BAL_T_cells[, BAL_T_cells$cluster %in% c("6", "7", "10")]
PBMC_T_cells <- corrected_PBMC[rownames(corrected_PBMC) %in% T_cell_genes, ]
PBMC_T_cells <- PBMC_T_cells[, PBMC_T_cells$seurat_clusters %in% c("4", "21", "28")]
rowData(BAL_T_cells)$scmap_features <- NULL
rowData(PBMC_T_cells)$scmap_features <- NULL
colData(BAL_T_cells)$cluster <- NULL
colData(PBMC_T_cells)$seurat_clusters <- NULL
Combined_T_Cells <- cbind(BAL_T_cells, PBMC_T_cells)

p2 <- plotDots(Combined_T_Cells, features=rownames(Combined_T_Cells), group="verbose_cluster") + 
  theme(aspect.ratio = 4, axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.title.x=element_blank(), axis.title.y=element_blank())

png(filename='Results/Bubbleplot_T_and_B.png', height=2160, width=540)
grid.arrange(p1, p2, ncol=1)
dev.off()

## Everything bubbleplot
#reorder axis's using heatmap dendrogram
pp <- heatmap.2(heat_all_Matrix, scale="row")
rowIndex <- pp$rowInd
colIndex <- pp$colInd

sce1 <- corrected_BAL
sce2 <- corrected_PBMC
rowData(sce1)$scmap_features <- NULL
rowData(sce2)$scmap_features <- NULL
colData(sce1)$cluster <- NULL
colData(sce2)$seurat_clusters <- NULL
Combined_sce <- cbind(sce1, sce2)

png(filename='Results/bubbleplot_scmap_all.png', height=3840, width=3840)
plotDots(Combined_sce, features=rownames(heat_all_Matrix), group="verbose_cluster", color=turbo(256)) +
  theme_minimal(base_size = 50) +
  theme(aspect.ratio = 1, axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.title.x=element_blank(), axis.title.y=element_blank(), legend.key.size = unit(2, 'cm'), axis.text.y=element_text(size=10)) +
  scale_y_discrete(limits=rownames(heat_all_Matrix)[rowIndex]) +
  scale_x_discrete(limits=colnames(heat_all_Matrix)[colIndex])
dev.off()

```

# Wrapping up

Save everything

```{r save-everything}
sessionInfo()
save(list=ls(), file='Data/Validation_vs_PBMC_Save.RData') 
```